x-pack/platform/plugins/private/reporting/server/routes/common/request_handler/validator.ts

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import moment from 'moment-timezone';
+import { z } from '@kbn/zod';
+import DOMPurify from 'dompurify';
+import { JSDOM } from 'jsdom';
+import type { BaseParams } from '@kbn/reporting-common/types';
+
+const sanitizeString = (input: string) => {
+  if (typeof input !== 'string') {
+    return '';
+  }
+  const window = new JSDOM('').window;
+  const purify = DOMPurify(window);
+  return purify.sanitize(input);
+};
+
+export function validateTimezone(timezone: string) {
+  if (moment.tz.zone(timezone) != null) {
+    return;
+  }
+
+  return 'string is not a valid timezone: ' + timezone;
+}
+
+const timezoneSchema = z.string().refine((val) => validateTimezone(val) === undefined, {
+  message: 'Invalid timezone',
+});
+
+const dimensionsSchema = z
+  .object({
+    height: z.number().positive().max(14400),
+    width: z.number().positive().max(14400),
+  })
+  .strict();
+
+export const idSchema = z.enum(['preserve_layout', 'print', 'canvas', 'png']); // "png" is used in some tests, could be a legacy value
+
+const layoutSchema = z
+  .object({
+    id: idSchema.optional(),
+    dimensions: dimensionsSchema.optional(),
+    zoom: z.number().positive().max(1000).optional(),
+    selectors: z
+      .object({
+        screenshot: z.string().max(1024).optional(),
+        renderComplete: z.string().max(1024).optional(),
+        renderError: z.string().max(1024).optional(),
+        renderErrorAttribute: z.string().max(1024).optional(),
+        itemsCountAttribute: z.string().max(1024).optional(),
+        timefilterDurationAttribute: z.string().max(1024).optional(),
+      })
+      .strict()
+      .optional(),
+  })
+  .strict();
+
+const objectTypeSchema = z
+  .string()
+  .max(256)
+  .transform((objectType) => sanitizeString(objectType));
+
+const titleSchema = z
+  .string()
+  .max(2048)
+  .transform((title) => sanitizeString(title));
+const versionSchema = z
+  .string()
+  .max(32)
+  .transform((version) => sanitizeString(version));
+
+const forceNowSchema = z
+  .string()
+  .max(32)
+  .transform((forceNow) => sanitizeString(forceNow));
+
+export const pagingStrategySchema = z.enum(['pit', 'scroll']);
+
+const locatorObjectSchema = z.object({
+  id: z.string().max(1024).optional(),
+  version: z.string().max(32).optional(),
+  params: z.record(z.any()).optional(),
+});
+
+const locatorParamsSchema = z.array(locatorObjectSchema).max(1).or(locatorObjectSchema);
+
+const relativeUrlSchema = z
+  .string()
+  .max(4096)
+  .transform((relativeUrl) => sanitizeString(relativeUrl));
```

Comment:
would it also remove `&` from URLs?

Reply:
no it encodes it to `&amp;` added a test for it.